tealdbg command dispatch, debugging port 9392 configuration and --listen option auto setting

emg110 opened the pull request:
- Added port 9392 in docker-compose configuration file for algod section
- Added a "Debugging" section at the end of README.md
- Reflected the port availability and small tips to use it, in debugging section of README.md

winder commented:
@ryanRfox could you try this out?

emg110 commented:
What does Team Carbon-11 stands for if I might ask?

winder commented:
@emg110 We have a few feature teams, the one I'm on is Team Carbon-11. It's an element with a "fast" half-life.

emg110 commented:
The carbon11 and it's decay to Broron11 I knew but what the label means and pointed to was the question which I got the answer for. Thanks @winder .

yaovi-a commented:
@ryanRfox any comments on this?

emg110 commented:
Hi @ryanRfox , Hi @winder , Hi @yaovi-a  
Any updates on this PR?
I appreciate your time and kind care in advance

ryanRfox commented:
Thanks very much @emg110 for this PR. I got it to work as implemented. However, my preference is for some additional refactoring. 

As a (typical) Sandbox user I primarily use it as a _blackbox_ or an infrastructure component. I'm willing to issue `sandbox copy` and `sandbox goal` commands, but if I can avoid it, I'd rather not have to do `sandbox enter algod` then issue the `tealdbg` command and later `exit`.

Please, consider adding a `sandbox tealdbg` command which silently adds the required `--listen 0.0.0.0` flag. This way a dApp dev may deploy a their new app and debug the call transactions something like this:

```bash
sandbox copy approval.teal
sandbox copy clear.teal
sandbox goal app create --creator $CREATOR --local-byteslices 0 --global-byteslices 1 --local-ints 1 --global-ints 1 --approval-prog approval.teal --clear-prog clear.teal
sandbox goal app call --dryrun-dump --out dump.json --app-id $APP_ID --from $USER
sandbox tealdbg debug approval.teal --dryrun-req dump.json
```

Then from their Chrome browser `chrome://inspect` and connect to the debugger.

emg110 commented:
Thanks for review and detailed instructions @ryanRfox , Consider it done! I got the point and will make your proposed adjustments ASAP.

emg110 commented:
Hi dear @ryanRfox , the changes are done including:

- Automatic proper configuration of --listener and prompting user about it
- Updated README.md on file works examples.
- updated README.md on tealdbg usage.
- updated the PR to include commits ahead from main sandbox repository and therefore it's up to date and the merge will be smooth.

emg110 commented:
Hi guys, changes are done and the travis CI build fail is probably because of one of two commits the main branch was ahead and I merged them before sending my new commits.
Build error is not mine and has some problem with nightly config:
![image](https://user-images.githubusercontent.com/1900448/126878456-d9785e0f-f503-44d3-a772-b38415be669a.png)

emg110 commented:
Any updates?

winder commented on README.md:
```diff
@@ -137,6 +142,22 @@ export INDEXER_SHA=""
+
+Algorand smart contract debugging process uses `tealdbg` command line of algod instance(algod container in sandbox).
+
+***Debug examples:
+
+**** Debugging smart contract with Chrome Developer Tools (CDT):  `tealdbg debug ${TEAL_PROGRAM} -f cdt -d dryrun.json`.
+
+**** Debugging smart contract with Web Interface (primal web ui) `tealdbg debug ${TEAL_PROGRAM} -f web -d dryrun.json`.
```
```suggestion
#### tealdbg examples:

Debugging smart contract with Chrome Developer Tools (CDT):
```~$ ./sandbox tealdbg debug ${TEAL_PROGRAM} -f cdt -d dryrun.json```

Debugging smart contract with Web Interface (primal web UI)
```~$ ./sandbox tealdbg debug ${TEAL_PROGRAM} -f web -d dryrun.json```
```

winder commented:
@emg110 thanks, the doc updates look very nice. Ryan is on vacation, once he gets back and trys this out we should be able to merge it in.

emg110 commented:
Thanks @winder , appreciate very good suggestions and committed them.

ryanRfox commented:
Yes! works as I desired. Thanks much @emg110

ryanRfox commented on README.md:
```diff
@@ -23,9 +23,10 @@ sandbox commands:
+  logs            -> stream algorand logs with the carpenter utility.
+  status          -> get node status.
+  goal (args)     -> run goal command like 'goal node status'.
+  tealdbg (args)  -> run tealdbg command like 'tealdbg help'. In sandbox, always use `tealdbg` with `--listen 0.0.0.0` option!
```
Change the description to:
"-> run tealdbg command to debug program execution."

You can drop the listen flag notice as it is appended by default.

emg110 replied:
Good, I changed the sentence as instructed by @ryanRfox .
In case of --listen flag, I have added this because default appended --listen flag is set to 127.0.0.1 as of listening on localhost loopback network only and that reminder notifies users to change it to 0.0.0.0 in case for listening on all container attached networks or specific container IP address that user prefers. Now considering this fact, I'm awaiting further instructions!

ryanRfox reviewed (changes requested):
Minor update requested to the command description.

emg110 commented:
Good, I changed the sentence as instructed by @ryanRfox .
In case of `--listen` flag, I have added this because default appended --listen flag is set to `127.0.0.1` as of listening on localhost loopback network only and that reminder notifies users to change it to `0.0.0.0` in case for listening on all container attached networks or specific container IP address that user prefers. Now considering this fact, I'm awaiting further instructions!

ryanRfox commented on README.md:
```diff
@@ -23,9 +23,10 @@ sandbox commands:
+  logs            -> stream algorand logs with the carpenter utility.
+  status          -> get node status.
+  goal (args)     -> run goal command like 'goal node status'.
+  tealdbg (args)  -> run tealdbg command to debug program execution. Always use `tealdbg` with `--listen 0.0.0.0` or `--listen [IP ADDRESS]` falgs, if you need to acess tealdbg from outside of algod docker container.
```
Please relocate this to the README.md file:
"Always use `tealdbg` with `--listen 0.0.0.0` or `--listen [IP ADDRESS]` falgs, if you need to acess tealdbg from outside of algod docker container."

ryanRfox commented on README.md:
```diff
@@ -137,6 +142,24 @@ export INDEXER_SHA=""
+For detailed information on how to debug smart contracts and use tealdbg CLI,please consult with [Algorand Development Portal :: Smart Contract Debugging](https://developer.algorand.org/docs/features/asc1/debugging/#setting-the-debugger-context).
+
+Algorand smart contract debugging process uses `tealdbg` command line of algod instance(algod container in sandbox).
+
```
Please add the --listen flag comments in this area.

emg110 replied:
Right away Sir!

emg110 replied:
It is done!

ryanRfox reviewed (changes requested):
Small relocation request.
